libs/wingc/src/lib.rs

```diff
@@ -48,6 +48,50 @@ pub struct CompilerOutput {
 	pub diagnostics: Diagnostics,
 }
 
+// Hack needed for some WASM related reason
+#[cfg(target_arch = "wasm32")]
+#[no_mangle]
+pub extern "C" fn _start() -> i32 {
+	0
+}
+
+unsafe fn ptr_to_string(ptr: u32, len: u32) -> String {
+	let slice = std::slice::from_raw_parts(ptr as *const u8, len as usize);
+	String::from_utf8_unchecked(slice.to_vec())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wingc_malloc(size: u32) -> *mut u8 {
+	let layout = core::alloc::Layout::from_size_align_unchecked(size as usize, 0);
+	std::alloc::alloc(layout)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wingc_free(ptr: u32, size: u32) {
+	let layout = core::alloc::Layout::from_size_align_unchecked(size as usize, 0);
+	std::alloc::dealloc(ptr as *mut u8, layout);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wingc_compile(ptr: u32, len: u32) {
+	let args = ptr_to_string(ptr, len);
+
+	let split = args.split(";").collect::<Vec<&str>>();
+	let source_file = split[0];
+	let output_dir = split.get(1).map(|s| *s);
+
+	let results = compile(source_file, output_dir);
+	if let Err(mut err) = results {
+		// Sort error messages by line number (ascending)
+		err.sort_by(|a, b| a.cmp(&b));
+		eprintln!(
+			"Compilation failed with {} errors\n{}",
+			err.len(),
+			err.iter().map(|d| format!("{}", d)).collect::<Vec<_>>().join("\n")
+		);
+	}
```

Comment:
Can we output these errors in some kind of structured format (eg. JSON) so that the responsibility of error formatting etc. can be deferred to whoever is consuming wingc? Right now some error formatting is done in the wingc (like this snippet), and some in the CLI (eg. #1034), and I feel like if we're treating wingc like a library, it will be easier to maintain if we keep all of the formatting outside.

Reply:
Totally agreed, but I think it'd be okay to leave that for a follow-up (I'll make an issue). In this PR I'm pretty much exclusively changing how wingc is invoked and otherwise changing as little else as possible.